### DISCUSSION TITLE
How use: `bind -m vi-insert -x '"\C-r": __fzf_history__'` in clifm?

### DISCUSSION THREAD

**Discussion by LinArcX** [leo-arch/clifm#282] [Q&A] [answered]
Hi. One thing that I still miss when I work with Clifm is the __Ctrl-r__ functionality to search through command history interactively. the default __Ctrl-r__ in bash is not my favorite. I like [this approach](https://github.com/junegunn/fzf/blob/master/shell/key-bindings.bash#L48) and I used it a lot in the past.

I was going to implement it on my own in a simpler way like this:(`clih.sh`)
`cat ~/.bash_history | fzf --prompt="bash history: " --height=12 --cycle --tac`

The only problem is that whenever I select one of the items from the list and press enter, it will paste it to stdout and run it. but I don't like this behavior, I want to paste it to stdout and __stay there__!

If you look at that file [at the end,](https://github.com/junegunn/fzf/blob/master/shell/key-bindings.bash#L126) they solved it by using `bind` functionality. How we can do it in Clifm? 

Edit: it's good to mention that I put `clih.sh` into plugins directory and assigned it like this in actions: `plugin2=clih.sh`
and also in keybindings file: `plugin2:\C-r`. So keybindings are working perfectly!

**Comment by leo-arch** [comment, marked as answer]
Hi @LinArcX.

Though it does not provide exactly the same functionality, you can try with `!<TAB>` (the main difference being that you get **clifm**'s command history, not bash).

EDIT: Since the latest commit it is possible to bind `!<TAB>` to a custom keybinding via the `cmd-hist` action in the keybindings file. For example, to bind it to <kbd>Ctrl-r</kbd>:

1. Edit the keybindings file: `kb edit`
2. Add this: `cmd-hist:\C-r`.
3. Restart Clifm.

**Reply by LinArcX** [reply]
> Since the latest commit it is possible to bind !<TAB> to a custom keybinding via the cmd-hist action in the keybindings file. 

Ah, cool. awesome. thanks a lot.

It's working now. the only thing that still can be improved is that we can sort the list items based on the latest usage of them. in this way, I can see them quickly at the top of the list, navigate to them, and select them. no need to type.

**Reply by leo-arch** [reply]
> the only thing that still can be improved is that we can sort the list items based on the latest usage of them

Recall that, if you want the last executed command, you can still just press the Up arrow key, just as you do in other shells.

**Reply by LinArcX** [reply]
i know. the default sorting is not what i want.
